host/class/uac/usb_host_uac/uac_host.c

```diff
@@ -0,0 +1,2141 @@
+    }
+}
+
+/**
+ * @brief UAC Host release Interface and free transfers, change state to IDLE
+ *
+ * @param[in] iface       Pointer to Interface structure,
+ * @return esp_err_t
+ */
+static esp_err_t uac_host_interface_release_and_free_transfer(uac_iface_t *iface)
+{
+    UAC_RETURN_ON_INVALID_ARG(iface);
+    UAC_RETURN_ON_INVALID_ARG(iface->parent);
+
+    UAC_RETURN_ON_FALSE(is_interface_in_list(iface), ESP_ERR_NOT_FOUND, "Interface handle not found");
+    UAC_RETURN_ON_ERROR(usb_host_interface_release(s_uac_driver->client_handle, iface->parent->dev_hdl, iface->dev_info.iface_num), "Unable to release UAC Interface");
+
+    if (iface->free_xfer_list) {
+        for (int i = 0; i < iface->xfer_num; i++) {
+            if (iface->free_xfer_list[i]) {
+                ESP_ERROR_CHECK(usb_host_transfer_free(iface->free_xfer_list[i]));
+            }
+        }
+        free(iface->free_xfer_list);
+    }
+
+    if (iface->xfer_list) {
+        for (int i = 0; i < iface->xfer_num; i++) {
+            if (iface->xfer_list[i]) {
+                ESP_ERROR_CHECK(usb_host_transfer_free(iface->xfer_list[i]));
+            }
+        }
+        free(iface->xfer_list);
+    }
+
+    // Change state
+    iface->state = UAC_INTERFACE_STATE_IDLE;
+    return ESP_OK;
+}
+
+/**
+ * @brief UAC Host claim Interface and prepare transfer, change state to READY
+ *
+ * @param[in] iface       Pointer to Interface structure,
+ * @return esp_err_t
+ */
+static esp_err_t uac_host_interface_claim_and_prepare_transfer(uac_iface_t *iface)
+{
+    esp_err_t ret = ESP_OK;
+    // Claim Interface
+
+    UAC_RETURN_ON_ERROR(usb_host_interface_claim(s_uac_driver->client_handle, iface->parent->dev_hdl, iface->dev_info.iface_num,
+                        iface->cur_alt + 1), "Unable to claim Interface");
+    // alloc a list of usb transfer
+    uint32_t packet_size = iface->iface_alt[iface->cur_alt].ep_mps;
+    iface->xfer_list = calloc(iface->xfer_num, sizeof(usb_transfer_t *));
+    UAC_GOTO_ON_FALSE(iface->xfer_list, ESP_ERR_NO_MEM, "Unable to allocate transfer list");
+    iface->free_xfer_list = calloc(iface->xfer_num, sizeof(usb_transfer_t *));
+    UAC_GOTO_ON_FALSE(iface->free_xfer_list, ESP_ERR_NO_MEM, "Unable to allocate free transfer list");
+    for (int i = 0; i < iface->xfer_num; i++) {
+        UAC_GOTO_ON_ERROR(usb_host_transfer_alloc(packet_size * iface->packet_num, iface->packet_num, &iface->free_xfer_list[i]),
+                          "Unable to allocate transfer buffer for EP IN");
+    }
+    // Change state
+    iface->state = UAC_INTERFACE_STATE_READY;
+    return ESP_OK;
+
+fail:
+    uac_host_interface_release_and_free_transfer(iface);
+    return ret;
+}
+
+/**
+ * @brief UAC IN Transfer complete callback
+ *
+ * @param[in] transfer  Pointer to transfer data structure
+ */
+static void stream_rx_xfer_done(usb_transfer_t *in_xfer)
+{
+    assert(in_xfer);
+
+    uac_iface_t *iface = get_interface_by_ep(in_xfer->bEndpointAddress);
```

Comment:
Here we take the interface by EP number (Yes, I know that I did that and it is a cp from HID).

Now I know, that it can't be simplified: 
1. Currently, `context = iface->parent` when then there is only the comparison in line 931 and do not use it later in `stream_rx_xfer_done`.
2. We can use this `context` and point it to `iface` directly. And then, during the `xfer_done` callback, we should only verify that `context` is not `NULL`. If it is not `NULL`, we have the interface pointer. 

In that case, it could help us to eliminate `get_interface_by_ep` and simplify `stream_rx_xfer_done` logic.

Reply:
@roma-jam make sense, thanks!